### PR TITLE
UCT/CUDA/BASE: Get device for current context only if context is active.

### PR DESCRIPTION
**Pull request by rakhmets**
## What
Added condition to call `cuCtxGetDevice` only if there is a CUDA context bound to the calling CPU thread. Otherwise, `cuDeviceGet(ordinal = 0)` is used in query method.

## Why ?
Previously `cuCtxGetDevice` could return `CUDA_ERROR_INVALID_CONTEXT` error. The error was not traced, but could be caught by `compute-sanitaizer`.

### CODE REVIEW

**Review by brminich — DISMISSED** [review]

File: src/uct/cuda/base/cuda_iface.c [hunk 1]
```diff
@@ -39,8 +39,13 @@ uct_cuda_base_query_devices_common(
...
     CUdevice cuda_device;
 
-    if (cuCtxGetDevice(&cuda_device) == CUDA_SUCCESS) {
+    if (uct_cuda_base_is_context_active()) {
+        if (UCT_CUDADRV_FUNC_LOG_ERR(cuCtxGetDevice(&cuda_device)) != UCS_OK) {
+            return UCS_ERR_NO_DEVICE;
+        }
         uct_cuda_base_get_sys_dev(cuda_device, &sys_device);
+    } else {
+        ucs_debug("querying devices without active context");
```
Comment (yosefe) [inline src/uct/cuda/base/cuda_iface.c:48]: this means we won't know the sys device of the GPU? it can affect protocols and devices selection..

Comment (rakhmets) [reply src/uct/cuda/base/cuda_iface.c:48]: Yes. And this is current behavior. `cuCtxGetDevice` returns `CUDA_ERROR_INVALID_CONTEXT`, if there is no active device.
We can try to add here `cuDeviceGet(&cuda_device, 0)`, assuming that 0-th device will be used...

Comment (yosefe) [reply src/uct/cuda/base/cuda_iface.c:48]: Yes, let's do this, and also update the debug message to print that no device is selected and we're taking locality information from device 0

File: src/uct/cuda/base/cuda_iface.c [hunk 2]
```diff
@@ -36,13 +36,23 @@ uct_cuda_base_query_devices_common(
...
+    ucs_sys_device_t sys_device;
     CUdevice cuda_device;
+    ucs_status_t status;
 
-    if (cuCtxGetDevice(&cuda_device) == CUDA_SUCCESS) {
-        uct_cuda_base_get_sys_dev(cuda_device, &sys_device);
+    if (uct_cuda_base_is_context_active()) {
+        status = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxGetDevice(&cuda_device));
+    } else {
+        ucs_debug("no device is selected, zero device will be used");
```
[the remark this answers is not in the document]
Comment (rakhmets) [reply src/uct/cuda/base/cuda_iface.c]: Fixed.

**Review by yosefe — APPROVED** [review]

File: src/uct/cuda/base/cuda_iface.c (hunk 1 above)
Comment (rakhmets) [reply src/uct/cuda/base/cuda_iface.c]: Reverted this change because of wire incompatibility issue. https://github.com/openucx/ucx/pull/9529

**Pull request merged**